utils/thapi_gen_source_callbacks.rb

```diff
@@ -0,0 +1,205 @@
+# callbacks SOURCE type plugin that will generate the same event stream.
+#
+# For use in creating test cases for unit testing, with input trace files that
+# are easy to maintain.
+#
+# Does not yet handle complex struct args, so the output may not be identical when
+# roundtripping through babeltrace_thapi.
+
+require 'optparse'
+require 'yaml'
+require 'erb'
+require 'time'
+require 'set'
+
+SOURCE_TEMPLATE = <<~TEXT.freeze
+  /* Code generated by #{__FILE__} */
+
+  #include <babeltrace2/babeltrace.h>
+  #include <metababel/metababel.h>
+
+  static const void** EMPTY_ARRAY;
+
+  void btx_push_usr_messages(void *btx_handle, void *usr_data, btx_source_status_t *status) {
+      <%- if not data.empty? and not data.first[:hostname].nil? -%>
+      btx_downstream_set_environment_hostname(btx_handle, "<%= data.first[:hostname] %>");
+      <%- end -%>
+
+      <%- data.each do | entry | -%>
+      <%- entry.fetch(:times,1).times do -%>
+      btx_push_message_<%= entry[:name] %>(btx_handle<%= ', ' if not entry[:field_values].empty? %><%= 
+       entry[:field_casts].zip(entry[:field_values].map{ |v| v.to_s }).map{ |cv| '(' + cv[0] + ')' + cv[1] }.join(', ')
+      %>);
+      <%- end -%>
+      <%- end -%>
+
+      *status = BTX_SOURCE_END;
+  }
+
+  void btx_register_usr_callbacks(void *btx_handle) {
+    btx_register_callbacks_push_usr_messages(btx_handle, &btx_push_usr_messages);
+  }
+TEXT
+
+BASE_DATE = Time.utc(2024)
+
+def get_values(name, value)
+  if not name.end_with?("_vals") or not value.strip.start_with?("[")
+    # simple value, do not modify
+    return [value]
+  end
+
+  # value will look like '[ ]' or '[ value ]' or '[ val1, val2, ... ]'
+  count = 0
+  value = value.strip[1...-1].strip
+  if value.empty?
+    count = 0
+    # use static const empty array defined in callback template
+    value = "EMPTY_ARRAY";
+  else
+    values = value.split(/, */)
+    count = 1
+    # TODO: handle multiple values
+    value = values[0]
+  end
+  [count, value]
+  # hack: always return empty array
+  [0, "EMPTY_ARRAY"]
+end
+
+def parse_field_values(fields_string, exclude_fields)
+  # expects commas separated list of name: value pairs
+  values = []
+  fields_string.split(/, */).each do |nv_pair|
+    name, value = nv_pair.split(/: /)
+    name.strip!
+    if not exclude_fields.member?(name)
+      values += get_values(name, value)
+    end
+  end
+  values
+end
+
+def get_fields(model, event_name, stream_class=0)
+  fields = {}
+  stream_classes = model[:stream_classes]
+  sc = stream_classes[0]
+  sc[:event_common_context_field_class][:members].each do |field_info|
+    fields[field_info[:name]] = field_info[:field_class][:cast_type]
+  end
+  sc[:event_classes].each do |eclass|
+    if eclass[:name] != event_name
+      next
+    end
+    eclass[:payload_field_class][:members].each do |field_info|
+      fields[field_info[:name]] = field_info[:field_class][:cast_type]
+    end
+    break
+  end
```

Comment:
```suggestion
  
  eclass = sc[:event_classes].find { |eclass| eclass[:name] == event_name}
 field_class = eclass[:members].to_h |field_info|
     [ field_info[:name], field_info[:field_class][:cast_type] ]
 end
 fields.update(field_class)
```
(and same idea for the first `each`. Let try to not mutate variable :)

Reply:
I see the value of find, but using to_h and .update seems more complicated than just setting fields[key] = v directly once the matching eclass is found

Reply:
I guess isolates the non-functional part to the end, each is kind of ugly.

Reply:
I hate mutating variables. Make the code hard to flow (add non-local behavior).  

At least here, it's clear that the `update` will overwrite the keys.

Removing the `break` and the `next`, to remove any kind of control flow is important. the `to_h` less :)

Reply:
ok I am warming to to_h, still getting used to the block syntax